core/src/main/java/io/grpc/Context.java

```diff
@@ -0,0 +1,496 @@
+package io.grpc;
+
+import com.google.common.base.Preconditions;
+import com.google.common.util.concurrent.MoreExecutors;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+
+import javax.annotation.Nullable;
+
+
+/**
+ * --Add warning about using this too liberally--
+ * --Add warning about using this with containers that unload Classloaders--
+ * -- Add note that in the case of cascading closure that listeners to the parent are always
+ * notified before listeners on the children.
+ *
+ */
+public class Context {
```

Comment:
Just a discussion point: do we want this exposed at the top-level of the API in `io.grpc`? It doesn't seem like something that all users would necessarily use and isn't really part of the API in general. Would it make more sense in a subpackage?